wdl/transforms/new-base/src/main/scala/wdl/transforms/base/wdlom2wom/expression/WdlomWomExpression.scala

```diff
@@ -36,9 +35,24 @@ final case class WdlomWomExpression private (expressionElement: ExpressionElemen
   // NB types can be determined using the linked values, so we don't need the inputMap:
   override def evaluateType(inputMap: Map[String, WomType]): ErrorOr[WomType] = evaluatedType
 
-  override def evaluateFiles(inputs: Map[String, WomValue], ioFunctionSet: IoFunctionSet, coerceTo: WomType): ErrorOr[Set[FileEvaluation]] =
-    // If WDL ever supports optional outputs then the `optional` parameter of `FileEvaluation` would need to be assigned appropriately.
-    expressionElement.evaluateFilesNeededToEvaluate(inputs, ioFunctionSet, coerceTo) map { _ map FileEvaluation.requiredFile }
+  /** Returns `true` if the specified `WomType` contains only optional file types. If the specified `WomType` is
+    * complex and we can't work out definitively whether all its file types are optional, return `false` to be on the
+    * safe side. */
+  def areAllFileTypesInWomTypeOptional(womType: WomType): Boolean = womType match {
+    case WomOptionalType(_: WomPrimitiveFileType) => true
+    case _: WomPrimitiveFileType => false
+    case _: WomPrimitiveType => true // WomPairTypes may have a non-File component here, this is fine.
+    case WomArrayType(inner) => areAllFileTypesInWomTypeOptional(inner)
+    case WomMapType(_, inner) => areAllFileTypesInWomTypeOptional(inner)
+    case WomPairType(leftType, rightType) => areAllFileTypesInWomTypeOptional(leftType) && areAllFileTypesInWomTypeOptional(rightType)
+    case _ => false
+  }
+
+  override def evaluateFiles(inputs: Map[String, WomValue], ioFunctionSet: IoFunctionSet, coerceTo: WomType): ErrorOr[Set[FileEvaluation]] = {
+    expressionElement.evaluateFilesNeededToEvaluate(inputs, ioFunctionSet, coerceTo) map { _ map {
+      FileEvaluation(_, optional = areAllFileTypesInWomTypeOptional(coerceTo), secondary = false)
```

Comment:
I guess ideally the `evaluateFilesNeededToEvaluate` would return `FileEvaluation`s natively rather than having to map in this all-or-nothing way. Infinitely better than the current state of the world with slight potential for risk in edge cases like:
```
Struct foo {
  File bam
  File? bai
}
```

Reply:
Or `Pair[File, File?]`, I suppose

Reply:
But happy to leave this as a TODO after the release and/or if anyone cares about it

Reply:
Yes exactly. Given the current state of file evaluation I'm going to make this cautious and only mark files as optional for unambiguous cases.

Reply:
oh btw call caching appears to be broken with missing optional output files so I'll need to make a ticket for that too 🤕